### PR TITLE
Implement the estimate gradient function for the FIR filter

### LINKED ISSUE #992: LFilter and LFilterPyTorch provide inconsistent gradients
**Describe the bug**
LFilter and LFilterPyTorch provide inconsistent gradients. In particular, the LFilter ones are incorrect.

**To Reproduce**
```
import numpy as np
import torch
from torch.autograd import Variable
from art.preprocessing.audio import LFilter, LFilterPyTorch

f = LFilter(numerator_coef=np.array([1.0, 0.0, 1.0]), denominator_coef=np.array([1.0, 0.0, 0.0]))
f_py = LFilterPyTorch(numerator_coef=np.array([1.0, 0.0, 1.0]), denominator_coef=np.array([1.0, 0.0, 0.0]))
x = Variable(torch.randn(10), requires_grad=True)

f.estimate_gradient(x.detach().numpy(), grad=np.ones(10))
# array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

f_py.estimate_gradient(x.detach().numpy(), grad=np.ones(10))
# array([2., 2., 2., 2., 2., 2., 2., 2., 1., 1.], dtype=float32)

# This estimate is equivalent to:
z, _ = f_py.forward(x)
z.backward(torch.ones(10))
x.grad
```

**Expected behavior**
LFilter is using identity BPDA to estimate the gradient, but this is incorrect. I'm not sure how much work it is to implement for IIR filters, but FIR should be pretty straightforward.

**System information (please complete the following information):**
ART version 1.6.0

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
This pr implements the estimate gradient function for the FIR filter numpy version.

Fixes #992 

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/preprocessing/audio/l_filter/numpy.py [hunk 1]
```diff
@@ -108,6 +108,87 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.nd
...
+
+        :param x: Input data for which the gradient is estimated. First dimension is the batch size.
+        :param grad: Gradient value so far.
+        :return: The gradient (estimate) of the defence.
+        """
+        if (grad.shape != x.shape) and (
+            (len(grad.shape) != len(x.shape) + 1) or (grad.shape[2:] != x.shape[1:]) or (grad.shape[0] != x.shape[0])
+        ):
+            raise ValueError(
+                "The `so far` gradient shape {} does not suit the input shape {}".format(grad.shape, x.shape)
```
Comment (beat-buesser) [inline art/preprocessing/audio/l_filter/numpy.py:123]:
```suggestion
                "The shape of `grad` {} does not match the shape of input `x` {}".format(grad.shape, x.shape)
```

File: art/preprocessing/audio/l_filter/numpy.py [hunk 2]
```diff
@@ -108,6 +108,87 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.nd
...
+            )
+
+        if not (self.denominator_coef[0] == 1.0 and np.sum(self.denominator_coef) == 1.0):
```
Comment (beat-buesser) [inline art/preprocessing/audio/l_filter/numpy.py:126]:
```suggestion
        if self.denominator_coef[0] != 1.0 and np.sum(self.denominator_coef) != 1.0:
```

File: art/preprocessing/audio/l_filter/numpy.py [hunk 3]
```diff
@@ -108,6 +108,87 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.nd
...
+            logger.warning(
+                "Gradient estimation for a non-fir filtering operation is not supported. A BPDA gradient estimation "
+                "is returned."
```
Comment (beat-buesser) [inline art/preprocessing/audio/l_filter/numpy.py:129]:
```suggestion
                "Accurate gradient estimation is currently only implemented for finite impulse response filtering, the coefficients indicate infinite response filtering, therefore Backward Pass Differentiable Approximation (BPDA) is applied instead."
```

File: art/preprocessing/audio/l_filter/numpy.py [hunk 4]
```diff
@@ -108,6 +108,87 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.nd
...
+                # First compute the gradient matrix
+                grad_matrix = self._compute_gradient_matrix(size=x_i.size)
+
+                # Then combine with the input gradient
+                grad_x_i = np.zeros_like(grad[i])
+                for j in range(x_i.size):
+                    grad_x_i[:, j] = np.sum(grad[i, :, :] * grad_matrix[:, j], axis=1)
+
+                # And store the result
+                x_grad_list.append(grad_x_i)
```
Comment (beat-buesser) [inline art/preprocessing/audio/l_filter/numpy.py:160]: The two outer for-loops are very similar except of their inner for-loops. Could we combine the two outer for-loops by moving the if-condition into the loop and reduce the number of identical lines?

Comment (minhitbk) [reply art/preprocessing/audio/l_filter/numpy.py:159]: @beat-buesser : Yes I thought about this, then I decided this form because I don't want the if-condition to be called repeatedly and unnecessarily in the for loop. But it will a bit slower but that's ok to move the if-condition inside.

**Review by beat-buesser — APPROVED** [review]

**Comment by beat-buesser** [comment]
@minhitbk Thank you!
